### ISSUE TITLE
Crash in legalize input types

### ISSUE THREAD

**Issue by GMNGeoffrey** [google/iree#6227]
https://gist.github.com/GMNGeoffrey/c1da910992cedaaf92ce3184e77af6f1 has a repro that can be run with iree-tf-opt

> assert.h assertion failed at llvm-project/mlir/lib/IR/Block.cpp:258 in mlir::Operation *mlir::B
lock::getTerminator(): !empty() && back().mightHaveTrait<OpTrait::IsTerminator>()

This is hit when running a transformer model through iree-import-xla

**Comment by GMNGeoffrey** [comment]
Scratch that, iree-tf-opt won't process it but iree-opt will (even though it's part of the import pipeline) :-/

**Comment by GMNGeoffrey** [comment]
The failure comes from:

https://github.com/google/iree/blob/303f865bbaafed4b73532987a6eb8f8bbab1590f/iree/compiler/Dialect/Flow/Transforms/LegalizeInputTypes.cpp#L124

Seems like having the verifier itself crash is pretty unhelpful...

**Comment by GMNGeoffrey** [comment]
Through the magic of printf debugging, I think the new op triggering the assert is

```mlir
%1010:2 = "mhlo.sort"(%1008, %1009) ( {                                                                                                                       
^bb0(%arg66: tensor<f32>, %arg67: tensor<f32>, %arg68: tensor<i32>, %arg69: tensor<i32>):  // no predecessors                                                 
  %1014 = "mhlo.bitcast_convert"(%arg66) : (tensor<f32>) -> tensor<i32>                                                                                       
  %1015 = "mhlo.compare"(%1014, %26) {comparison_direction = "LT"} : (tensor<i32>, tensor<i32>) -> tensor<i1>                                                 
  %1016 = "mhlo.bitcast_convert"(%arg66) : (tensor<f32>) -> tensor<ui32>                                                                                      
  %1017 = mhlo.subtract %57, %1016 : tensor<ui32>                                                                                                             
  %1018 = "mhlo.bitcast_convert"(%1017) : (tensor<ui32>) -> tensor<i32>                                                                                       
  %1019 = "mhlo.select"(%1015, %1018, %1014) : (tensor<i1>, tensor<i32>, tensor<i32>) -> tensor<i32>                                                          
  %1020 = "mhlo.bitcast_convert"(%arg67) : (tensor<f32>) -> tensor<i32>                                                                                       
  %1021 = "mhlo.compare"(%1020, %26) {comparison_direction = "LT"} : (tensor<i32>, tensor<i32>) -> tensor<i1>                                                 
  %1022 = "mhlo.bitcast_convert"(%arg67) : (tensor<f32>) -> tensor<ui32>                                                                                      
  %1023 = mhlo.subtract %57, %1022 : tensor<ui32>                                                                                                             
  %1024 = "mhlo.bitcast_convert"(%1023) : (tensor<ui32>) -> tensor<i32>                                                                                       
  %1025 = "mhlo.select"(%1021, %1024, %1020) : (tensor<i1>, tensor<i32>, tensor<i32>) -> tensor<i32>                                                          
  %1026 = "mhlo.compare"(%1019, %1025) {comparison_direction = "LT"} : (tensor<i32>, tensor<i32>) -> tensor<i1>                                               
  "mhlo.return"(%1026) : (tensor<i1>) -> ()                                                                                                                   
}) {dimension = 1 : i64, is_stable = true} : (tensor<1x134832xf32>, tensor<1x134832xi32>) -> (tensor<1x134832xf32>, tensor<1x134832xi32>)  
```

I don't understand though because that sure looks like a terminator to me. Maybe something isn't registered right?

**Comment by GMNGeoffrey** [comment]
Ok if I just disable verification of the new op in there, the reproducer then passes. Which is kind of weird because verify should be running as a separate pass afterwards. Regardless I shouldn't be able to crash the verifier here...

**Comment by GMNGeoffrey** [comment]
Yeah I can't figure out exactly what's going on, but by deleting all the sort ops I get this pass to run without error

**Comment by GMNGeoffrey** [comment]
But extracting the sort ops into a separate function doesn't reproduce the issue

**Comment by GMNGeoffrey** [comment]
If I get it to dump immediately before asserting. The dump it gives is actually for the entire func op. Here's a func op that crashes the verifier: https://gist.github.com/GMNGeoffrey/877336064cb6398f7b9b6acf5de080ea

**Comment by GMNGeoffrey** [comment]
Ok, a much smaller repro

```mlir
func @main(%arg666 : tensor<1x134832xf32>, %arg667 : tensor<1x134832xi32>), %arg668 : tensor<i1>) -> tensor<1x7xi32> {
  cond_br %arg666, ^bb2, ^bb3
^bb2:  // pred: ^bb1
  %1010:2 = "mhlo.sort"(%arg666, %arg667) ( {
  ^bb0:  // no predecessors
    "mhlo.return"(%arg668) : (tensor<i1>) -> ()
  }) {dimension = 1 : i64, is_stable = true} : (tensor<1x134832xf32>, tensor<1x134832xi32>) -> (tensor<1x134832xf32>, tensor<1x134832xi32>)
^bb3:  // pred: ^bb1
}
```

**Comment by GMNGeoffrey** [comment]
Ok a much smaller repro for real this time

```mlir
func @main(%arg666 : i1, %arg667 : tensor<1x134832xf32>, %arg668 : tensor<1x134832xi32>, %arg669 : tensor<i1>) {
  cond_br %arg666, ^bb2, ^bb3
^bb2:  // pred: ^bb1
  %1010:2 = "mhlo.sort"(%arg667, %arg668) ( {
  ^bb0(%arg66: tensor<f32>, %arg67: tensor<f32>, %arg68: tensor<i32>, %arg69: tensor<i32>):  // no predecessors
    "mhlo.return"(%arg669) : (tensor<i1>) -> ()
  }) {dimension = 1 : i64, is_stable = true} : (tensor<1x134832xf32>, tensor<1x134832xi32>) -> (tensor<1x134832xf32>, tensor<1x134832xi32>)
^bb3:  // pred: ^bb1
}
```

**Comment by GMNGeoffrey** [comment]
Somehow some part of verification is detecting mhlo.sort as maybe a terminator at one point and definitely not a terminator at another point. I wonder if registration is somehow happening part way through verification??

```mlir
func @main(%arg666 : i1, %arg667 : tensor<1x134832xf32>, %arg668 : tensor<1x134832xi32>, %arg669 : tensor<i1>) {
  br ^bb2
^bb2:  // pred: ^bb1
  %1010:2 = "mhlo.sort"(%arg667, %arg668) ( {
  ^bb0(%arg66: tensor<f32>, %arg67: tensor<f32>, %arg68: tensor<i32>, %arg69: tensor<i32>):  // no predecessors
    "mhlo.return"(%arg669) : (tensor<i1>) -> ()
  }) {dimension = 1 : i64, is_stable = true} : (tensor<1x134832xf32>, tensor<1x134832xi32>) -> (tensor<1x134832xf32>, tensor<1x134832xi32>)
}
```

**Comment by GMNGeoffrey** [comment]
I tried setting preloadDialectsInContext to true to see if it was an issue with late loading of the MHLO dialect, but that doesn't seem to have fixed it

**Comment by GMNGeoffrey** [comment]
This has been identified as a bug in the MLIR verifier. I'll just quote @joker-eph:

> I think I found the broken underlying assumption. 
When the verifier starts on a module or a function, it'll first verify the nested region before looking for the ops inside the region.
But when you invoke the verifier on a non-isolated-from-above operation directly, it hasn't verified whatever is in the parent regions. But when it looks for SSA value, it'll check for dominance where these values are defined.
> 
> And dominance assumes that we first check the region.

Basically when trying to verify the sort op (which is not isolated from above) the function it's in is not complete yet. The verifier tries to verify the SSA values used in the sort op are dominated by it, but dominance checking relies on the function being well formed (specifically it calls getTerminator).

**Comment by joker-eph** [comment]
Should be fixed with https://github.com/llvm/llvm-project/commit/a6559b42cee2
